File: go-sdk/kestra_api_client/worker_groups_api.go
package kestra_api_client

import "context"

// WorkerGroupsAPI covers the instance-level /api/v1/instance/worker-groups
// routes that are usable on a plain EE instance: reads and token management.
// The create/update/delete and subscription endpoints are feature-gated
// (FEATURE_WORKER_GROUP / FEATURE_WORKER_QUEUE) and are not wrapped here.
type WorkerGroupsAPI struct {
	baseAPI
}

// ListWorkerGroups returns every worker group. The optional filters use the
// WORKER_GROUP query filter format.
func (a *WorkerGroupsAPI) ListWorkerGroups(ctx context.Context, filters []SearchFilter) (*ApiWorkerGroupList, error) {
	params := buildQueryParams()
	appendFilterParams(params, filters)
	return doJSON[*ApiWorkerGroupList](&a.baseAPI, ctx, "GET", superadminPath("instance", "worker-groups"), nil, params)
}

// WorkerGroup returns a single worker group by id.
func (a *WorkerGroupsAPI) WorkerGroup(ctx context.Context, id string) (*ApiWorkerGroup, error) {
	return doJSON[*ApiWorkerGroup](&a.baseAPI, ctx, "GET", superadminPath("instance", "worker-groups", id), nil, nil)
}

// CreateWorkerGroup creates a worker group. Requires the WORKER_GROUP feature,
// and the license caps how many worker groups may exist.
func (a *WorkerGroupsAPI) CreateWorkerGroup(ctx context.Context, request ApiCreateWorkerGroupRequest) (*ApiWorkerGroup, error) {
	return doJSON[*ApiWorkerGroup](&a.baseAPI, ctx, "POST", superadminPath("instance", "worker-groups"), request, nil)
}

// UpdateWorkerGroup updates the worker group with the given id.
func (a *WorkerGroupsAPI) UpdateWorkerGroup(ctx context.Context, id string, request ApiUpdateWorkerGroupRequest) (*ApiWorkerGroup, error) {
	return doJSON[*ApiWorkerGroup](&a.baseAPI, ctx, "PUT", superadminPath("instance", "worker-groups", id), request, nil)
}

// DeleteWorkerGroup deletes the worker group with the given id. force removes it
// even if workers are still active. The default group cannot be deleted (400).
func (a *WorkerGroupsAPI) DeleteWorkerGroup(ctx context.Context, id string, force *bool) error {
	params := buildQueryParams("force", force)
	return a.doVoid(ctx, "DELETE", superadminPath("instance", "worker-groups", id), nil, params)
}

// AddWorkerGroupSubscription subscribes the worker group to a worker queue.
// Requires the WORKER_QUEUE feature and an existing worker queue.
func (a *WorkerGroupsAPI) AddWorkerGroupSubscription(ctx context.Context, id string, request ApiSubscriptionRequest) (*ApiWorkerGroup, error) {
	return doJSON[*ApiWorkerGroup](&a.baseAPI, ctx, "POST", superadminPath("instance", "worker-groups", id, "subscriptions"), request, nil)
}

// UpdateWorkerGroupSubscription changes the reserved capacity of a subscription.
func (a *WorkerGroupsAPI) UpdateWorkerGroupSubscription(ctx context.Context, id, workerQueueId string, request ApiUpdateSubscriptionRequest) (*ApiWorkerGroup, error) {
	return doJSON[*ApiWorkerGroup](&a.baseAPI, ctx, "PATCH", superadminPath("instance", "worker-groups", id, "subscriptions", workerQueueId), request, nil)
}

// RemoveWorkerGroupSubscription unsubscribes the worker group from a worker queue.
func (a *WorkerGroupsAPI) RemoveWorkerGroupSubscription(ctx context.Context, id, workerQueueId string) (*ApiWorkerGroup, error) {
	return doJSON[*ApiWorkerGroup](&a.baseAPI, ctx, "DELETE", superadminPath("instance", "worker-groups", id, "subscriptions", workerQueueId), nil, nil)
}

// GenerateWorkerGroupToken mints a registration token for a worker group. The
// plaintext token is returned exactly once, in the response.
func (a *WorkerGroupsAPI) GenerateWorkerGroupToken(ctx context.Context, id string, request ApiGenerateTokenRequest) (*ApiGenerateTokenResponse, error) {
	return doJSON[*ApiGenerateTokenResponse](&a.baseAPI, ctx, "POST", superadminPath("instance", "worker-groups", id, "tokens"), request, nil)
}

// RevokeWorkerGroupToken revokes a worker group's registration token without
// deleting it; the returned group reflects the token's revoked state.
func (a *WorkerGroupsAPI) RevokeWorkerGroupToken(ctx context.Context, id, tokenId string) (*ApiWorkerGroup, error) {
	return doJSON[*ApiWorkerGroup](&a.baseAPI, ctx, "POST", superadminPath("instance", "worker-groups", id, "tokens", tokenId, "revoke"), nil, nil)
}

// DeleteWorkerGroupToken permanently deletes a worker group's registration token.
func (a *WorkerGroupsAPI) DeleteWorkerGroupToken(ctx context.Context, id, tokenId string) error {
	return a.doVoid(ctx, "DELETE", superadminPath("instance", "worker-groups", id, "tokens", tokenId), nil, nil)
}

File: python/python-sdk/kestrapy/api/worker_groups_api.py
from typing import Any

from kestrapy.base_api import BaseApi
from kestrapy.models.worker_group_controller_api_create_worker_group_request import WorkerGroupControllerApiCreateWorkerGroupRequest
from kestrapy.models.worker_group_controller_api_generate_token_request import WorkerGroupControllerApiGenerateTokenRequest
from kestrapy.models.worker_group_controller_api_generate_token_response import WorkerGroupControllerApiGenerateTokenResponse
from kestrapy.models.worker_group_controller_api_subscription_request import WorkerGroupControllerApiSubscriptionRequest
from kestrapy.models.worker_group_controller_api_update_subscription_request import WorkerGroupControllerApiUpdateSubscriptionRequest
from kestrapy.models.worker_group_controller_api_update_worker_group_request import WorkerGroupControllerApiUpdateWorkerGroupRequest
from kestrapy.models.worker_group_controller_api_worker_group import WorkerGroupControllerApiWorkerGroup
from kestrapy.models.worker_group_controller_api_worker_group_capacity import WorkerGroupControllerApiWorkerGroupCapacity
from kestrapy.models.worker_group_controller_api_worker_group_list import WorkerGroupControllerApiWorkerGroupList
from kestrapy.models.worker_group_controller_api_worker_group_worker_list import WorkerGroupControllerApiWorkerGroupWorkerList


class WorkerGroupsApi(BaseApi):
    """Instance-scoped worker groups (`/api/v1/instance/worker-groups`). Instance-owner-only."""


    # ---- CRUD ----

    def list_worker_groups(self) -> WorkerGroupControllerApiWorkerGroupList:
        path = self._superadmin_path("instance", "worker-groups")
        return self._json_request("GET", path, WorkerGroupControllerApiWorkerGroupList)

    def worker_group(self, id: str) -> WorkerGroupControllerApiWorkerGroup:
        path = self._superadmin_path("instance", "worker-groups", id)
        return self._json_request("GET", path, WorkerGroupControllerApiWorkerGroup)

    def create_worker_group(self, request: WorkerGroupControllerApiCreateWorkerGroupRequest) -> WorkerGroupControllerApiWorkerGroup:
        path = self._superadmin_path("instance", "worker-groups")
        return self._json_request("POST", path, WorkerGroupControllerApiWorkerGroup, body=request)

    def update_worker_group(self, id: str, request: WorkerGroupControllerApiUpdateWorkerGroupRequest) -> WorkerGroupControllerApiWorkerGroup:
        path = self._superadmin_path("instance", "worker-groups", id)
        return self._json_request("PUT", path, WorkerGroupControllerApiWorkerGroup, body=request)

    def delete_worker_group(self, id: str) -> None:
        path = self._superadmin_path("instance", "worker-groups", id)
        self._void_request("DELETE", path)

    # ---- Subscriptions ----

    def add_worker_group_subscription(self, id: str, request: WorkerGroupControllerApiSubscriptionRequest) -> WorkerGroupControllerApiWorkerGroup:
        path = self._superadmin_path("instance", "worker-groups", id, "subscriptions")
        return self._json_request("POST", path, WorkerGroupControllerApiWorkerGroup, body=request)

    def update_worker_group_subscription_reservation(self, id: str, worker_queue_id: str, request: WorkerGroupControllerApiUpdateSubscriptionRequest) -> WorkerGroupControllerApiWorkerGroup:
        path = self._superadmin_path("instance", "worker-groups", id, "subscriptions", worker_queue_id)
        return self._json_request("PATCH", path, WorkerGroupControllerApiWorkerGroup, body=request, content_type=self.JSON)

    def remove_worker_group_subscription(self, id: str, worker_queue_id: str) -> WorkerGroupControllerApiWorkerGroup:
        path = self._superadmin_path("instance", "worker-groups", id, "subscriptions", worker_queue_id)
        return self._json_request("DELETE", path, WorkerGroupControllerApiWorkerGroup)

    # ---- Tokens ----

    def generate_worker_group_token(self, id: str, request: WorkerGroupControllerApiGenerateTokenRequest) -> WorkerGroupControllerApiGenerateTokenResponse:
        path = self._superadmin_path("instance", "worker-groups", id, "tokens")
        return self._json_request("POST", path, WorkerGroupControllerApiGenerateTokenResponse, body=request)

    def revoke_worker_group_token(self, id: str, token_id: str) -> WorkerGroupControllerApiWorkerGroup:
        path = self._superadmin_path("instance", "worker-groups", id, "tokens", token_id, "revoke")
        return self._json_request("POST", path, WorkerGroupControllerApiWorkerGroup)

    def delete_worker_group_token(self, id: str, token_id: str) -> None:
        path = self._superadmin_path("instance", "worker-groups", id, "tokens", token_id)
        self._void_request("DELETE", path)

    # ---- Runtime ----

    def list_worker_group_workers(self, id: str) -> WorkerGroupControllerApiWorkerGroupWorkerList:
        path = self._superadmin_path("instance", "worker-groups", id, "workers")
        return self._json_request("GET", path, WorkerGroupControllerApiWorkerGroupWorkerList)

    def worker_group_capacity(self, id: str) -> WorkerGroupControllerApiWorkerGroupCapacity:
        path = self._superadmin_path("instance", "worker-groups", id, "capacity")
        return self._json_request("GET", path, WorkerGroupControllerApiWorkerGroupCapacity)
